### PR TITLE
Add pre-commit to check that aiobotocore is an optional dependency

### PR DESCRIPTION
**Pull request by potiuk**
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
image

**Comment by potiuk** [comment]
We do not want to check it manually, we alredy got it accidentally merged once for the last amazon's provider's release And we could have easily repeat the mistake: https://github.com/apache/airflow/pull/31018

File: scripts/ci/pre_commit/pre_commit_check_aiobotocore_optional.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        console.print(f"[blue]Checking[/blue]: {file}")
+        provider_yaml_content = yaml.safe_load(Path(file).read_text())
+        dependencies = provider_yaml_content.get("dependencies")
+        if dependencies and any(dependency.startswith("aiobotocore") for dependency in dependencies):
+            errors.append(
+                f"\n[red]Error: the aibotocore cannot be a required dependency, "
+                f"because it restricts botocore too much[/]\n\n"
+                f"The [magenta]{file}[/] file has aiobotocore dependency set at top level.\n\n"
+                f"[yellow]Please remove it and make sure it is added only as"
+                f"an optional dependency in additional-extras\n"
```
Comment (pankajkoti) [inline scripts/ci/pre_commit/pre_commit_check_aiobotocore_optional.py]:
```suggestion
                f"an optional dependency in additional-extras[/]\n"
```

Safer to close [yellow]?

File: scripts/ci/pre_commit/pre_commit_check_aiobotocore_optional.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    for file in sys.argv[1:]:
+        console.print(f"[blue]Checking[/blue]: {file}")
+        provider_yaml_content = yaml.safe_load(Path(file).read_text())
+        dependencies = provider_yaml_content.get("dependencies")
+        if dependencies and any(dependency.startswith("aiobotocore") for dependency in dependencies):
+            errors.append(
+                f"\n[red]Error: the aibotocore cannot be a required dependency, "
+                f"because it restricts botocore too much[/]\n\n"
+                f"The [magenta]{file}[/] file has aiobotocore dependency set at top level.\n\n"
+                f"[yellow]Please remove it and make sure it is added only as"
```
Comment (pankajkoti) [inline scripts/ci/pre_commit/pre_commit_check_aiobotocore_optional.py]:
```suggestion
                f"[yellow]Please remove it and make sure it is added only as "
```

**Review by pankajkoti — APPROVED** [review]

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -603,6 +603,14 @@ repos:
         language: python
         files: ^.*LICENSE.*$|^.*LICENCE.*$
         pass_filenames: false
+      - id: check-aiobotocore-optional
+        name: Check if aiobotocore is an optional dependency only
+        entry: ./scripts/ci/pre_commit/pre_commit_check_aiobotocore_optional.py
+        language: python
+        files: ^airflow/providers/.*/provider\.yaml$
```
Comment (eladkal) [inline .pre-commit-config.yaml:617]: It should run only on the amazon provider yaml does it not?

Comment (pankajkoti) [reply .pre-commit-config.yaml:617]: I had the same thought but then wondered that we have some transfer operators too like `s3_to_gcs` in the google providers for example and felt we might need this check there too 🤔

Comment (potiuk) [reply .pre-commit-config.yaml:617]: That was exactly my point - it costs us almost nothing to exclude aiobotocore from all providers.

File: scripts/ci/pre_commit/pre_commit_check_aiobotocore_optional.py (hunk 1 above)
Comment (potiuk) [reply scripts/ci/pre_commit/pre_commit_check_aiobotocore_optional.py]: wHy not :)

File: scripts/ci/pre_commit/pre_commit_check_aiobotocore_optional.py (hunk 2 above)
Comment (potiuk) [reply scripts/ci/pre_commit/pre_commit_check_aiobotocore_optional.py]: Right

**Review by ferruzzi — APPROVED** [review]
I also like that you added all providers, just to protect us against Transfers accidentally including it.  :+1:

**Review by o-nikolas — APPROVED** [review]

**Pull request merged**